File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/binary_tree/P654_MaximumBinaryTree.java
package com.dsa.leetcode.labuladong.binary_tree;

public class P654_MaximumBinaryTree {
    public TreeNode constructMaximumBinaryTree(int[] nums) {
        return build(nums, 0, nums.length - 1);
    }

    private TreeNode build(int[] nums, int left, int right) {
        if (left > right) {
            return null;
        }
        int maxIndex = left;
        int maxValue = nums[left];
        for (int i = left + 1; i <= right; i++) {
            if (maxValue < nums[i]) {
                maxIndex = i;
                maxValue = nums[i];
            }
        }

        TreeNode root = new TreeNode(maxValue);
        root.left = build(nums, left, maxIndex - 1);
        root.right = build(nums, maxIndex + 1, right);
        return root;
    }
}

File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong/binary_tree/P654_MaximumBinaryTree.py
# author: tobebetter9527
# since: 2024/2/28 8:42
from typing import List, Optional

from com.dsa.leetcode.labuladong.binary_tree import TreeNode


class Solution:
    def constructMaximumBinaryTree(self, nums: List[int]) -> Optional[TreeNode]:
        return self.build(nums, 0, len(nums) - 1)

    def build(self, nums, left, right):
        if left > right:
            return None
        max_index, max_value = left, nums[left]
        for i in range(left + 1, right + 1):
            if max_value < nums[i]:
                max_index = i
                max_value = nums[i]
        root = TreeNode(max_value)
        root.left = self.build(nums, left, max_index - 1)
        root.right = self.build(nums, max_index + 1, right)
        return root
